### PR TITLE
Add workingdir integration test

### LINKED ISSUE #1624: Add missing integration test for workingdir
Following review comment from @ImJasonH here : 

https://github.com/tektoncd/pipeline/pull/1609#pullrequestreview-321907838

Let's add an integration test for the workingdir feature

/area testing
/assign
/kind feature

### PR DESCRIPTION
**Pull request by chmouel**
Add an integration test checking the workingdirs

Following on review https://github.com/tektoncd/pipeline/pull/1609#pullrequestreview-321907838 we realised that there wasn't any integration test for the workingdir step, let's add some tests for this.

Closes #1624

# Changes

We didn't check previously if workingdir really worked properly, let's add tests
for this.

/cc @ImJasonH 

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [X] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [X] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

### CODE REVIEW

**Comment by chmouel** [comment]
I could not find a way to do a `TestBad` I need to figure out how to not being able to write to /workspace since it's volume mounted, fs mode rights are assigned to our running user (so having a new SA with non elevated priv would not work.

**Comment by ImJasonH** [comment]
Would it be possible to phrase at least some of this test as a YAML test?

```
steps:
- image: ubuntu
  command: ['touch', 'foo']
  workingDir: 'foo'  # implicitly mkdir's /workspace/foo
- image: ubuntu
  command: ['cat', '/workspace/foo/foo']
...
```

This has the benefit of being easier (I think) to read and understand in isolation.

**Comment by chmouel** [comment]
Good idea, you meant something like updating the existing [workdir.yaml](https://github.com/tektoncd/pipeline/blob/master/examples/taskruns/workingdir.yaml) with : 

```diff
diff --git a/examples/taskruns/workingdir.yaml b/examples/taskruns/workingdir.yaml
index e8a7d87f..73229cf8 100644
--- a/examples/taskruns/workingdir.yaml
+++ b/examples/taskruns/workingdir.yaml
@@ -12,6 +12,14 @@ spec:
       - '-c'
       - '[[ $PWD == /workspace ]]'
 
+    - name: default
+      image: ubuntu
+      command: ['bash']
+      workingDir: "foo"  # implicitly mkdir's /workspace/foo
+      args:
+      - '-c'
+      - '[[ $PWD == /workspace/foo ]]'
+
     - name: override
       image: ubuntu
       command: ['bash']
```

**Comment by ImJasonH** [comment]
Yeah that sounds great. Is there anything else left in the Go test that can be YAMLified?

**Comment by chmouel** [comment]
@ImJasonH I think the go test are still useful they make it implicit  what should be expected, I 

I meant we need to make sure no output is coming back out of the `mkdir -p` and i don't thin kthe YAML would catch this?

There is another test in there too which make sure that workingdir with absolute path (ie `/foo`) that are not starting with `/workspace`.

**Comment by ImJasonH** [comment]
> @ImJasonH I think the go test are still useful they make it implicit what should be expected, I
> 
> I meant we need to make sure no output is coming back out of the `mkdir -p` and i don't thin kthe YAML would catch this?

It wouldn't. I'm just not sure how much that matters though. The test implicitly checks that the directory exists as expected, that's all I really care about.

**Comment by bobcatfish** [comment]
@chmouel could you add some more detail in the commit message about what this commit is for? at the moment it just says "Add workingdir integration test" but it's hard to understand without context. I don't know what "workingdir integration test" is, and even from https://github.com/tektoncd/pipeline/issues/1624 I'm not really sure what the workingdir feature is :S 

Sorry for the nitpicking!! But having a lot of info in our [commit messages](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages) and [in our issues](https://github.com/tektoncd/pipeline/blob/master/.github/issue-template.md) is really important to me, and makes it a lot easier to make sure all the contributors can be on the same page.

@chmouel (or anyone) feel free to remove this hold once we have some more details:
/hold

**Comment by chmouel** [comment]
@bobcatfish I kinda rephrased it, but would you like me to explain what the workingdir step is doing in the commit message? I think that would be a documentation update, isnt it ?

**Comment by chmouel** [comment]
CI error is unrelated to this commit and this https://github.com/tektoncd/pipeline/pull/1634

**Comment by bobcatfish** [comment]
Thanks for the update! :pray: 

> @bobcatfish I kinda rephrased it, but would you like me to explain what the workingdir step is doing in the commit message? I think that would be a documentation update, isnt it ?

I mean I would! I like to err on the side of being verbose, cuz it's really hard to know how much you'll know later - for example, say someone is looking at this commit message 2 years from now and the workingdir step is long gone but some of the lines of code remain, they'll have a really hard time understanding this change

**Comment by chmouel** [comment]
@bobcatfish I understand and if I could I would have link to the documentation from the commit message to explains what the workingdir but unfortunately it's not there :(

**Comment by chmouel** [comment]
It doesnt not seem that a relative path for workingdir can actually work 🤔 

`
35s         Warning   Failed      pod/workingdir-2hcbb-pod-366869   Error: container create failed: Cwd must be an absolute path
`

**Comment by vdemeester** [comment]
> It doesnt not seem that a relative path for workingdir can actually work thinking
> 
> `35s Warning Failed pod/workingdir-2hcbb-pod-366869 Error: container create failed: Cwd must be an absolute path`

Yeah that make sense :angel:

**Comment by chmouel** [comment]
I guess this code is not doing what expected then :

https://github.com/chmouel/tektoncd-pipeline/blob/34db59dd10f6407de6427992b2d2b04c4208027b/pkg/pod/workingdir_init.go#L60-L65

It should be 

```go
	if filepath.IsAbs(p) || strings.HasPrefix(p, "/workspace/") {
```

(currently it skips the absolute path like `/foo` and catch the non absolute path like `foo`)

**Comment by chmouel** [comment]
CB> I guess this code is not doing what expected then :

I'll work on a following PR to fix the code,

File: test/workingdir_test.go [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+	t.Logf("Waiting for TaskRun in namespace %s to finish successfully", namespace)
+	if err := WaitForTaskRunState(c, wdTaskRunName, TaskRunSucceed(wdTaskRunName), "TaskRunSuccess"); err != nil {
+		t.Errorf("Error waiting for TaskRun to finish successfully: %s", err)
+	}
+
+	tr, err := c.TaskRunClient.Get(wdTaskRunName, metav1.GetOptions{})
+	if err != nil {
+		t.Errorf("Error retrieving taskrun: %s", err)
+	}
+	if tr.Status.PodName != "" {
```
Comment (vdemeester) [inline test/workingdir_test.go]: Should we fail in case of `tr.Status.Podman` being empty ?

Comment (chmouel) [reply test/workingdir_test.go]: Sure why not 👍

**Pull request merged**